### PR TITLE
Laboratory work #2, Polina Kobyakova - 22FPL2

### PR DESCRIPTION
**Pull request by kobyakovapolina**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
👀

File: docs/main.py [hunk 1]
```diff
@@ -0,0 +1 @@
+print("hello world")
```
Comment (artyomtugaryov) [inline docs/main.py:1]: remove

File: lab_1_classify_by_unigrams/start.py [hunk 2]
```diff
@@ -34,5 +33,6 @@ def main() -> None:
 
     assert result, "Detection result is None"
 
+
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:36]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+def prepare_word():
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/__init__.py:1]: done

File: seminars/practice_3_lists.py [hunk 4]
```diff
@@ -13,27 +13,27 @@
 
 # Create a list
 example = [1, 2, 3]
-print(example)
+#print(example)
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply seminars/practice_3_lists.py:16]: done

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -26,20 +39,41 @@ def collect_frequencies(
...
+    raw_words = {
+        prepare_word(word, start_of_word, end_of_word): text.count(word_list[i]) for i, word in enumerate(word_list)
+    }
+
+    return raw_words
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/main.py:54]: done

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -14,10 +15,22 @@ def prepare_word(
...
+        new_word = new_word + tuple(end_of_word.split())
+    if start_of_word and not end_of_word:
+        new_word = tuple(start_of_word.split()) + new_word
+
+    return new_word
 
 
 def collect_frequencies(
-    text: str, start_of_word: str | None, end_of_word: str
+        text: str, start_of_word: str | None, end_of_word: str
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/main.py:33]: done

File: docs/main.py (hunk 1 above)
Comment (kobyakovapolina) [reply docs/main.py:1]: done

File: lab_1_classify_by_unigrams/start.py (hunk 2 above)
Comment (kobyakovapolina) [reply lab_1_classify_by_unigrams/start.py:36]: Why should I do a revert? My start.py is like in start.py from repository.

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -14,10 +15,22 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if not isinstance(raw_word, str) or not isinstance(start_of_word, str | None) or not isinstance(end_of_word, str):
+        return None
+
+    new_word = tuple(raw_word)
+    if start_of_word and end_of_word:
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/main.py:22]: fixed

File: docs/main.py (hunk 1 above)
Comment (artyomtugaryov) [reply docs/main.py:1]: I don't think so

File: lab_1_classify_by_unigrams/start.py (hunk 2 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/start.py:36]: what do you say about the 36 line?

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_2_tokenize_by_bpe/main.py:31]: no

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_2_tokenize_by_bpe/main.py:52]: no

File: lab_2_tokenize_by_bpe/main.py [hunk 8]
```diff
@@ -26,42 +37,88 @@ def collect_frequencies(
...
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+
+    merged_tokens_dict = {}
+
+    for word in word_frequencies:
+        word_list = list(word)
+        if pair[0] and pair[1] in word:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:90]: are you sure you need the if?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: docs/main.py (hunk 1 above)
Comment (kobyakovapolina) [reply docs/main.py]: I guess I fixed it

File: lab_2_tokenize_by_bpe/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/__init__.py]: Now I believe I fixed

File: seminars/practice_3_lists.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply seminars/practice_3_lists.py:16]: fixed?

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/main.py]: fixed?

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_2_tokenize_by_bpe/main.py]: fixed?

File: lab_2_tokenize_by_bpe/main.py [hunk 9]
```diff
@@ -47,6 +82,21 @@ def merge_tokens(
...
+        return None
+
+    merged_tokens_dict = {}
+
+    for word in word_frequencies:
+        word_list = list(word)
+        for i, new_word in enumerate(word[:-1]):
+            new_word = (word[i], word[i + 1])
+            if new_word == pair:
+                word_list[i + 1] = pair[0] + pair[1]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:95]: join

File: lab_2_tokenize_by_bpe/main.py [hunk 10]
```diff
@@ -105,11 +178,11 @@ def load_vocabulary(vocab_path: str) -> dict[str, int] | None:
 
 
 def encode(
-    original_text: str,
-    vocabulary: dict[str, int] | None,
-    start_of_word_token: str | None,
-    end_of_word_token: str | None,
-    unknown_token: str,
+        original_text: str,
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:181]: revert

File: lab_2_tokenize_by_bpe/main.py [hunk 11]
```diff
@@ -132,7 +205,7 @@ def collect_ngrams(text: str, order: int) -> list[tuple[str, ...]] | None:
 
 
 def calculate_precision(
-    actual: list[tuple[str, ...]], reference: list[tuple[str, ...]]
+        actual: list[tuple[str, ...]], reference: list[tuple[str, ...]]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:208]: revert

File: seminars/practice_3_lists.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply seminars/practice_3_lists.py:16]: Do you see changes in this file in this PR? I see. Looks like you shoud revert it!

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**